hal/armv7a/imx6ull/imx6ull.c

```diff
@@ -0,0 +1,360 @@
+
+	*(imx6ull_common.iomux + grp_addds) = 0x00000030;
+
+	*(imx6ull_common.iomux + pad_dram_reset) = 0x00000030;
+	*(imx6ull_common.iomux + pad_dram_sdba2) = 0x00000000;
+	*(imx6ull_common.iomux + pad_dram_odt0) = 0x00000030;
+	*(imx6ull_common.iomux + pad_dram_odt1) = 0x00000030;
+
+	*(imx6ull_common.iomux + grp_ctlds) = 0x00000030;
+	*(imx6ull_common.iomux + grp_ddrmode_ctl) = 0x00020000;
+
+	*(imx6ull_common.iomux + pad_dram_sdqs0p) = 0x00000030;
+	*(imx6ull_common.iomux + pad_dram_sdqs1p) = 0x00000030;
+
+	*(imx6ull_common.iomux + grp_ddrmode) = 0x00020000;
+	*(imx6ull_common.iomux + grp_b0ds) = 0x00000030;
+	*(imx6ull_common.iomux + grp_b1ds) = 0x00000030;
+
+	*(imx6ull_common.iomux + pad_dram_dqm0) = 0x00000030;
+	*(imx6ull_common.iomux + pad_dram_dqm1) = 0x00000030;
+
+	/* Config DDR control register */
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x00008000;
+	*(imx6ull_common.mmdc + mmdc_mpzqhwctrl) = 0xa1390003;
+	*(imx6ull_common.mmdc + mmdc_mpwldectrl0) = 0x00150019;
+	*(imx6ull_common.mmdc + mmdc_mpdgctrl0) = 0x41550153;
+	*(imx6ull_common.mmdc + mmdc_mprddlctl) = 0x40403a3e;
+	*(imx6ull_common.mmdc + mmdc_mpwrdlctl) = 0x40402f2a;
+
+	*(imx6ull_common.mmdc + mmdc_mprddqby0dl) = 0x33333333;
+	*(imx6ull_common.mmdc + mmdc_mprddqby1dl) = 0x33333333;
+
+	*(imx6ull_common.mmdc + mmdc_mpwrdqby0dl) = 0xf3333333;
+	*(imx6ull_common.mmdc + mmdc_mpwrdqby1dl) = 0xf3333333;
+
+	*(imx6ull_common.mmdc + mmdc_mpdccr) = 0x00944009;
+	*(imx6ull_common.mmdc + mmdc_mpmur0) = 0x00000800;
+
+	*(imx6ull_common.mmdc + mmdc_mdpdc) = 0x0002002d;
+	*(imx6ull_common.mmdc + mmdc_mdotc) = 0x1b333030;
+	*(imx6ull_common.mmdc + mmdc_mdcfg0) = 0x676b52f3;
+	*(imx6ull_common.mmdc + mmdc_mdcfg1) = 0xb66d0b63;
+	*(imx6ull_common.mmdc + mmdc_mdcfg2) = 0x01ff00db;
+	*(imx6ull_common.mmdc + mmdc_mdmisc) = 0x00201740;
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x00008000;
+	*(imx6ull_common.mmdc + mmdc_mdrwd) = 0x000026d2;
+	*(imx6ull_common.mmdc + mmdc_mdor) = 0x006b1023;
+	*(imx6ull_common.mmdc + mmdc_mdasp) = 0x00000047;
+
+	*(imx6ull_common.mmdc + mmdc_mdctl) = 0x82180000;
+	*(imx6ull_common.mmdc + mmdc_mppdcmpr2) = 0x00400000;
+
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x2008032;
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x8033;
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x48031;
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x15208030;
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x4008040;
+
+	*(imx6ull_common.mmdc + mmdc_mdref) = 0x800;
+
+	*(imx6ull_common.mmdc + mmdc_mpodtctrl) = 0x227;
+	*(imx6ull_common.mmdc + mmdc_mdpdc) = 0x2552d;
+	*(imx6ull_common.mmdc + mmdc_mapsr) = 0x11006;
+	*(imx6ull_common.mmdc + mmdc_mdscr) = 0x0;
+}
+
+
+static void imx6ull_pllInit(void)
+{
+	/* Set ARM clock to 792 MHz; set ARM clock divider to 1 */
+	*(imx6ull_common.ccm + ccm_cacrr) = 0;
+	hal_cpuDataSyncBarrier();
+
+	/* Enable USB2 PLL - 480 MHz */
+	*(imx6ull_common.ccm_analog + ccm_analog_pll_usb2) |= 0x3000;
+
+	/* Wait until USB2 PLL is unlocked */
```

Comment:
I think we are waiting for PLL to be locked (PLL is a closed loop system, it's output is stable when it's locked, not unlocked).

Reply:
I based on the code from kernel:
https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/def017c456d991e0546de388827e6437f4ddc9d7/hal/armv7a/imx6ull/_init.S#L390-L393

According to documentation: **1 - PLL is currently locked; 0 - PLL is not currently locked.**

Reply:
So this might be bug - we need PLL to be locked, only then it's initialization has been completed

Reply:
I have checked in data provided by the producer and you have right, in kernel's code is a bug. PLL should be locked.
For this reason PLLs haven't been set appropriate on `imx6ull`, we based on bootROM initial setup.

Due to the fact that our software based on this implementation I didn't make changes in PLL configuration. I have switch off this piece of code to be considered later.